datumaro/components/project.py

```diff
@@ -2,486 +2,2383 @@
+            # we can use data from the working directory.
+            obj_hash = current['config'].hash
+
+            data_dir = None
+            cached = False
+
+            target = ProjectBuildTargets.strip_target_name(current_name)
+            if self._tree.is_working_tree and target in self._tree.sources:
+                data_dir = self._project.source_data_dir(target)
+
+                wd_hash = working_dir_hashes.get(target)
+                if not wd_hash:
+                    if osp.isdir(data_dir):
+                        wd_hash = self._project.compute_source_hash(data_dir)
+                        working_dir_hashes[target] = wd_hash
+                    else:
+                        log.debug("Build: skipping checking working dir '%s', "
+                            "because it does not exist", data_dir)
+                        data_dir = None
+
+                if obj_hash != wd_hash:
+                    log.debug("Build: skipping loading stage '%s' from "
+                        "working dir '%s', because hashes does not match",
+                        current_name, data_dir)
+                    data_dir = None
+
+            if not data_dir:
+                if self._project._is_cached(obj_hash):
+                    data_dir = self._project.cache_path(obj_hash)
+                    cached = True
+                elif self._project._can_retrieve_from_vcs_cache(obj_hash):
+                    data_dir = self._project._materialize_obj(obj_hash)
+                    cached = True
+
+                if not data_dir or not osp.isdir(data_dir):
+                    log.debug("Build: skipping loading stage '%s' from "
+                        "cache obj '%s', because it is not available",
+                        current_name, obj_hash)
+                    return None
+
+            if data_dir:
+                assert osp.isdir(data_dir), data_dir
+                log.debug("Build: loading stage '%s' from '%s'",
+                    current_name, data_dir)
+                return ProjectSourceDataset(
+                    data_dir, self._tree, target, readonly=cached)
+
+            return None
+
+
+        graph = pipeline._graph
+        if len(graph) == 0:
+            raise EmptyPipelineError()
+
+        head = pipeline.head
+        if not head:
+            raise MissingPipelineHeadError()
+
+        # traverse the graph and initialize nodes from sources to the head
+        to_visit = [head]
+        while to_visit:
+            current_name = to_visit.pop()
+            current = graph.nodes[current_name]
+
+            assert current.get('dataset') is None
+
+            obj_hash = current['config'].hash
+            if obj_hash:
+                dataset = _try_load_from_cache(current_name, current)
+                if dataset is not None:
+                    current['dataset'] = dataset
+                    continue
+
+            uninitialized_parents = []
+            initialized_parents = []
```

Comment:
This variable seems unnecessary. It's only used after the following `if`, and you can only get there if either `graph.in_degree(current_name) == 0` or `uninitialized_parents`, and in both of those cases `initialized_parents` will be equal to `graph.predecessors(current_name)`. So it seems like you should be able to use `graph.predecessors` instead.

Reply:
It is also used in the assertion below, but there it probably can be changed to `in_degree()`

Reply:
Yes... and neither case contradicts what I said, since in both instances the variable contains all parents of the node.

Reply:
I changed, but it is a cosmetic change anyway.